src/main/java/info/tehnut/soulshards/api/BindingEvent.java

```diff
@@ -1,15 +1,54 @@
 package info.tehnut.soulshards.api;
 
-import net.fabricmc.fabric.util.HandlerArray;
+import net.fabricmc.fabric.api.event.Event;
+import net.fabricmc.fabric.api.event.EventFactory;
 import net.minecraft.entity.LivingEntity;
+import net.minecraft.util.ActionResult;
 import net.minecraft.util.Identifier;
 import net.minecraft.util.TypedActionResult;
 
 public class BindingEvent {
 
-    public static final HandlerArray<NewBinding> NEW_BINDING = new HandlerArray<>(NewBinding.class);
-    public static final HandlerArray<GainSouls> GAIN_SOULS = new HandlerArray<>(GainSouls.class);
-    public static final HandlerArray<GetEntityName> GET_ENTITY_NAME = new HandlerArray<>(GetEntityName.class);
+    public static final Event<NewBinding> NEW_BINDINGS = EventFactory.createArrayBacked(NewBinding.class,
+            (listeners) -> (entity, binding) -> {
+                TypedActionResult<IBinding> result = new TypedActionResult<>(ActionResult.PASS, binding);
+
+                for(NewBinding newBinding : listeners) {
+                    TypedActionResult<IBinding> currentResult = newBinding.onNewBinding(entity, binding);
+
+                    if(currentResult.getResult() != ActionResult.PASS) {
+                        result = currentResult;
+                    }
+                }
+
+                return result;
+            }
+	);
+
+    public static final Event<GainSouls> GAIN_SOULS = EventFactory.createArrayBacked(GainSouls.class,
+            (listeners) -> (entity, binding, amount) -> {
+                int soulsGained = amount;
+
+                for (GainSouls gainSouls : listeners) {
+                    int newSoulsGained = gainSouls.getGainedSouls(entity, binding, amount);
+                    if(newSoulsGained > 0) soulsGained = newSoulsGained;
```

Comment:
Is this working as intended ? Shouldn't it add the souls each time?

Reply:
My understanding was that it was "go until you find the first valid soul amount." Changed it with the latest commit.